### PR TITLE
range request support for share mode

### LINKED ISSUE #736: Support HTTP range requests for `wget --continue`
It's nice to be able to conveniently share private photos and videos over onionshare even over a shitty internet connection. At the moment I have to start 10 onionshare instances to share 100MB each because the connection is not reliable enough to share 1GB at a time, and onionshare doesn't support `torsocks wget --continue`.

All that is needed for this is for onionshare to support [HTTP Range requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests).

I see a similar proposal #175 was rejected, however I think you should reconsider. This one should be more simple than that to implement, but also importantly neither of these proposals would *require* the receiver to install extra software, they can still just use Tor Browser. The proposal only makes it more convenient for users that do have other tools like `wget` at their disposal, or a Firefox extension like `DownThemAll`.

### PR DESCRIPTION
**Pull request by heartsucker**
Fixes #736

### CODE REVIEW

File: onionshare/common.py [hunk 1]
```diff
@@ -18,20 +18,26 @@
...
 import os
 import platform
 import random
+import re
 import socket
 import sys
 import tempfile
 import threading
 import time
+from flask import abort
```
Comment (mig5) [inline onionshare/common.py]: I think I'd be OK with these functions being inside onionshare/web/share_mode.py so we don't have to import flask here, in other words keeping all this stuff specific to the Share Mode in the Web module. It looks like you are importing `flask.abort` already over there anyway.

We use 'Common' mainly for stuff that is used in both CLI and GUI mode, but Web is already common to both too and has all the flask logic.

Just my 2c though, will let @micahflee do a proper review :)

Thanks for contributing this, looking forward to testing soon!

**Comment by heartsucker** [comment]
Hey so I think this is ready for a second review. I tried to keep my diff small, but it's made those couple of functions a bit unwieldy. I can write some tests and then refactor after a merge perhaps?

**Comment by mig5** [comment]
I tried this PR but I couldn't get `torsocks wget --continue`, or pause/resume of download in Tor Browser, to work.

Each time I paused in Tor Browser or ctrl+c'd from CLI with `torsocks wget`, OnionShare picked up the fact and treated it as a Canceled download.

More work might be needed in the Web module to workaround a 'canceled' download that is communicated back to the GUI as such? Not sure...

Did you try this change in any particular way that worked for you?

**Comment by micahflee** [comment]
I'll try to review this some myself soon.

But I want to add (I could be wrong) but I'm pretty sure a `wget --continue` technically _is_ a new download. I think it checks how many bytes big the not-finished file you already have is, then makes a new GET request, but with a header saying to skip that many bytes before starting the download. Then it appends the download to to the end of the file you already had, instead of replacing it.

So I think, regardless, this is gonna be tricky to display in the UI. We have no way of knowing for sure that the user who is running `wget --continue` is the same user who canceled earlier. I think the canceled download is correct, but the new download needs convey that it's starting X bytes in. So like, if someone resumes a download that was 60% done, the new download should start the progress bar at 60%.

**Comment by heartsucker** [comment]
So I think I implemented this in an overly strict was (requiring the `if-Match` header, but that's not actually necessary).

As for cancelling, I think it would be better to say "incomplete" since we might not know *why* the DL stopped. Could be the user aborted or the connection popped. But that's another issue.

**Comment by heartsucker** [comment]
Ok, so it turns out my first impl missed all sorts of edge cases and also `curl` doesn't follow the RFC exactly (or so it seems?). So. I wrote a standalone lib for now to hammer out all the bugs in this, and then I'll merge it into here.

Take a peek maybe? https://github.com/heartsucker/flask-rangerequest

**Comment by heartsucker** [comment]
This works with `curl`, `wget`, and `http` but not with Cancel/Retry on TB, but only because I couldn't get TB to treat the connection as failed because I think the connection timeout is super high.

File: onionshare/web/share_mode.py [hunk 2]
```diff
@@ -124,112 +210,183 @@ def download_logic(slug_candidate=''):
...
+        if self.common.platform != 'Darwin':
+            sys.stdout.write("\n")
+
+        # Download is finished
+        if not self.web.stay_open:
+            self.download_in_progress = False
+
+        # Close the server, if necessary
+        if not self.web.stay_open and not canceled:
+            print(strings.strings)
```
Comment (micahflee) [inline onionshare/web/share_mode.py]: Oops.

**Comment by micahflee** [comment]
I love that the unit tests test with `wget`, `curl`, and `http`.

**Comment by micahflee** [comment]
This is looking really good, and I confirmed that range requests work with `wget --continue`. But also pause/resume isn't working in Tor Browser for me. Pause works, but when I resume the download is "Failed". Since the clients pretty much always use Tor Browser, I'd love to fix this. 

I think you're right about the Tor Browser connection failed timeout, but I'll spend a bit of time seeing if I can debug it.

**Comment by micahflee** [comment]
I managed to capture a traffic dump of what's going on with Firefox cancel/resume. I shared a 1.8gb file (an Ubuntu iso) using `--local-only`, and loaded URL (in this case http://127.0.0.1:17647/component-blazing) in Firefox. In the background, I was running wireshark sniffing traffic on my loopback device.

I started the download, then paused it, which showed up as cancelled in OnionShare. Then I clicked resume, and it showed up as "Failed" in Firefox. Here's the http conversation of the failed resume:

```
GET /component-blazing/download HTTP/1.1
Host: 127.0.0.1:17647
User-Agent: Mozilla/5.0 (X11; Fedora; Linux x86_64; rv:64.0) Gecko/20100101 Firefox/64.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
Connection: keep-alive
Range: bytes=158349073-
If-Match: "sha256:c15331e6f5f923579a3ed0ca8bc6aa2325c67293abdc875ba856eb5fac07705d"
If-Unmodified-Since: Sun, 06 Jan 2019 23:04:45 GMT

HTTP/1.0 304 NOT MODIFIED
Content-Disposition: attachment; filename=ubuntu-18.04.1-desktop-amd64.iso
Content-Security-Policy: default-src 'self'; style-src 'self'; script-src 'self'; img-src 'self' data:;
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Referrer-Policy: no-referrer
Server: OnionShare
Accept-Ranges: bytes
ETag: "sha256:c15331e6f5f923579a3ed0ca8bc6aa2325c67293abdc875ba856eb5fac07705d"
Vary: Accept-Encoding
Connection: close
Date: Sun, 06 Jan 2019 23:23:03 GMT
```

For comparison, I tried the same thing but with wget, and this time with a 291mb file (Debian iso). First I downloaded part of the file and pressed ctrl-c, then I ran `wget --continue` to finish downloading the file. Here's what the resume looks like:

```
GET /strangle-garland/download HTTP/1.1
Range: bytes=81408000-
User-Agent: Wget/1.19.5 (linux-gnu)
Accept: */*
Accept-Encoding: identity
Host: 127.0.0.1:17622
Connection: Keep-Alive

HTTP/1.0 206 PARTIAL CONTENT
Content-Type: application/octet-stream
Content-Length: 223727615
Content-Disposition: attachment; filename=debian-9.6.0-amd64-netinst.iso
Content-Security-Policy: default-src 'self'; style-src 'self'; script-src 'self'; img-src 'self' data:;
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Referrer-Policy: no-referrer
Server: OnionShare
Accept-Ranges: bytes
ETag: "sha256:c51d84019c3637ae9d12aa6658ea8c613860c776bd84c6a71eaaf765a0dd60fe"
Last-Modified: Sun, 06 Jan 2019 23:04:45 GMT
Vary: Accept-Encoding
Content-Range: bytes 81408000-305135615/305135616
Date: Sun, 06 Jan 2019 23:30:44 GMT
```

So it appears that the problem is that the OnionShare web server is returning `304 NOT MODIFIED` to Firefox, but `206 PARTIAL CONTENT` to wget.

Maybe this is happening because Firefox is sending these headers, while wget doesn't send them, and flask automatically deals with them?

```
If-Match: "sha256:c15331e6f5f923579a3ed0ca8bc6aa2325c67293abdc875ba856eb5fac07705d"
If-Unmodified-Since: Sun, 06 Jan 2019 23:04:45 GMT
```

For the `Range` header to work, it should maybe suppress those other headers somehow?

**Comment by heartsucker** [comment]
Ah dang, yeah I can see what the problem is. In the range/status code extraction, there's an error where the `If-Unmodified-Since` overrides the "decision" on how to handle the request. I'll write some additional tests to cover this in hopes of preventing regressions.

**Comment by micahflee** [comment]
Still doesn't work... I started to take a stab at debugging this myself, and while I made a little progress, I'm still mostly puzzled. I pushed my code into the `heartsucker-range-requests` [branch](https://github.com/micahflee/onionshare/tree/heartsucker-range-requests), which includes resolving the merge conflict.

First, I shared some files and started downloading it in Tor Browser. Here's what it looked like (with the headers expanded for readability):

```
[Jan 21 2019 04:12:35 PM] ShareModeWeb.download_logic: request headers:

Host: evsrh3qici7yygey.onion
User-Agent: Mozilla/5.0 (Windows NT 6.1; rv:60.0) Gecko/20100101 Firefox/60.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
Connection: keep-alive
Upgrade-Insecure-Requests: 1

[Jan 21 2019 04:12:35 PM] ShareModeWeb.download_logic: response range=(0, 13235121), status_code=200, headers=

Content-Type: application/zip
Content-Length: 13235121
Content-Disposition: attachment; filename=onionshare_xuoagi.zip
Content-Security-Policy: default-src 'self'; style-src 'self'; script-src 'self'; img-src 'self' data:;
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Referrer-Policy: no-referrer
Server: OnionShare
Accept-Ranges: bytes
ETag: "sha256:1d48df05c7b89426fed451a3e87b1c81fa0e3208dad0d5fbaaf64b9474e334e1"
Last-Modified: Tue, 22 Jan 2019 00:10:31 GMT
Vary: Accept-Encoding

127.0.0.1 - - [21/Jan/2019 16:12:35] "GET /yummy-creole/download HTTP/1.1" 200 -
#0: 100.0 KiB, 0.77%
#0: 200.0 KiB, 1.55%
```

Then paused the download in Tor Browser:

```
#0: 9.7 MiB, 76.60%
#0: 9.8 MiB, 77.37%
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/werkzeug/serving.py", line 261, in execute
    write(data)
  File "/usr/lib/python3/dist-packages/werkzeug/serving.py", line 242, in write
    self.wfile.write(data)
  File "/usr/lib/python3.6/socketserver.py", line 800, in write
    self._sock.sendall(b)
ConnectionResetError: [Errno 104] Connection reset by peer

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/user/code/onionshare/onionshare/web/share_mode.py", line 358, in generate
    yield chunk
GeneratorExit

[Jan 21 2019 04:12:43 PM] ShareModeWeb.generate: remote user canceled, it looks like
```

This is what it looks like to OnionShare when a user cancels a download -- basically, the socket just gets closed on the Tor Browser side. I catch this exception, but I'm displaying the stacktrace anyway. So far so good.

Then resume the download in Tor Browser:

```
[Jan 21 2019 04:13:12 PM] ShareModeWeb.download_logic: request headers:

Host: evsrh3qici7yygey.onion
User-Agent: Mozilla/5.0 (Windows NT 6.1; rv:60.0) Gecko/20100101 Firefox/60.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
Connection: keep-alive
Range: bytes=981596-
If-Match: "sha256:1d48df05c7b89426fed451a3e87b1c81fa0e3208dad0d5fbaaf64b9474e334e1"
If-Unmodified-Since: Tue, 22 Jan 2019 00:10:31 GMT

[Jan 21 2019 04:13:12 PM] ShareModeWeb.get_range_and_status_code: range header: bytes=981596-
[Jan 21 2019 04:13:12 PM] ShareModeWeb.get_range_and_status_code: if-range header: None
[Jan 21 2019 04:13:12 PM] ShareModeWeb.get_range_and_status_code: etag: "sha256:1d48df05c7b89426fed451a3e87b1c81fa0e3208dad0d5fbaaf64b9474e334e1"
[Jan 21 2019 04:13:12 PM] ShareModeWeb.download_logic: response range=(981596, 13235121), status_code=206, headers=

Content-Type: application/zip
Content-Length: 12253525
Content-Disposition: attachment; filename=onionshare_xuoagi.zip
Content-Security-Policy: default-src 'self'; style-src 'self'; script-src 'self'; img-src 'self' data:;
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Referrer-Policy: no-referrer
Server: OnionShare
Accept-Ranges: bytes
ETag: "sha256:1d48df05c7b89426fed451a3e87b1c81fa0e3208dad0d5fbaaf64b9474e334e1"
Last-Modified: Tue, 22 Jan 2019 00:10:31 GMT
Vary: Accept-Encoding
Content-Range: bytes 981596-13235121/13235122

127.0.0.1 - - [21/Jan/2019 16:13:12] "GET /yummy-creole/download HTTP/1.1" 206 -
#1: 2.0 MiB, 15.61%
#1: 2.1 MiB, 16.38%
#1: 2.2 MiB, 17.15%
#1: 2.3 MiB, 17.93%
#1: 2.4 MiB, 18.70%
...
#1: 8.9 MiB, 70.54%
#1: 9.0 MiB, 71.31%
#1: 9.1 MiB, 72.09%
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/werkzeug/serving.py", line 261, in execute
    write(data)
  File "/usr/lib/python3/dist-packages/werkzeug/serving.py", line 242, in write
    self.wfile.write(data)
  File "/usr/lib/python3.6/socketserver.py", line 800, in write
    self._sock.sendall(b)
ConnectionResetError: [Errno 104] Connection reset by peer

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/user/code/onionshare/onionshare/web/share_mode.py", line 358, in generate
    yield chunk
GeneratorExit

[Jan 21 2019 04:13:13 PM] ShareModeWeb.generate: remote user canceled, it looks like
```

It appears to start the download in the correct place, downloads a bit more for a second, and then fails, in the exact same way as if the user canceled the download (only I didn't cancel it). In Tor Browser it just says it failed.

Now, here I'm doing the same thing with wget. I started sharing some files, and then I downloaded them with `torify wget`:

```
[Jan 21 2019 04:17:13 PM] ShareModeWeb.download_logic: request headers:

User-Agent: Wget/1.19.5 (linux-gnu)
Accept: */*
Accept-Encoding: identity
Host: hrlzhqkrl6hfmlmb.onion
Connection: Keep-Alive

[Jan 21 2019 04:17:13 PM] ShareModeWeb.download_logic: response range=(0, 13235121), status_code=200, headers=

Content-Type: application/zip
Content-Length: 13235121
Content-Disposition: attachment; filename=onionshare_lgdg5f.zip
Content-Security-Policy: default-src 'self'; style-src 'self'; script-src 'self'; img-src 'self' data:;
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Referrer-Policy: no-referrer
Server: OnionShare
Accept-Ranges: bytes
ETag: "sha256:1d48df05c7b89426fed451a3e87b1c81fa0e3208dad0d5fbaaf64b9474e334e1"
Last-Modified: Tue, 22 Jan 2019 00:16:17 GMT
Vary: Accept-Encoding

127.0.0.1 - - [21/Jan/2019 16:17:13] "GET /untimely-pesky/download HTTP/1.1" 200 -
#0: 100.0 KiB, 0.77%
#0: 200.0 KiB, 1.55%
```

Then pressed ctrl-c, which cancels the download:

```
#0: 10.4 MiB, 82.79%
#0: 10.5 MiB, 83.56%
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/werkzeug/serving.py", line 261, in execute
    write(data)
  File "/usr/lib/python3/dist-packages/werkzeug/serving.py", line 242, in write
    self.wfile.write(data)
  File "/usr/lib/python3.6/socketserver.py", line 800, in write
    self._sock.sendall(b)
ConnectionResetError: [Errno 104] Connection reset by peer

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/user/code/onionshare/onionshare/web/share_mode.py", line 358, in generate
    yield chunk
GeneratorExit

[Jan 21 2019 04:17:19 PM] ShareModeWeb.generate: remote user canceled, it looks like
```

Then resume it with `torify wget --continue`:

```
[Jan 21 2019 04:18:31 PM] ShareModeWeb.download_logic: request headers:

Range: bytes=1247595-
User-Agent: Wget/1.19.5 (linux-gnu)
Accept: */*
Accept-Encoding: identity
Host: hrlzhqkrl6hfmlmb.onion
Connection: Keep-Alive

[Jan 21 2019 04:18:31 PM] ShareModeWeb.get_range_and_status_code: range header: bytes=1247595-
[Jan 21 2019 04:18:31 PM] ShareModeWeb.get_range_and_status_code: if-range header: None
[Jan 21 2019 04:18:31 PM] ShareModeWeb.get_range_and_status_code: etag: "sha256:1d48df05c7b89426fed451a3e87b1c81fa0e3208dad0d5fbaaf64b9474e334e1"
[Jan 21 2019 04:18:31 PM] ShareModeWeb.download_logic: response range=(1247595, 13235121), status_code=206, headers=

Content-Type: application/zip
Content-Length: 11987526
Content-Disposition: attachment; filename=onionshare_lgdg5f.zip
Content-Security-Policy: default-src 'self'; style-src 'self'; script-src 'self'; img-src 'self' data:;
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Referrer-Policy: no-referrer
Server: OnionShare
Accept-Ranges: bytes
ETag: "sha256:1d48df05c7b89426fed451a3e87b1c81fa0e3208dad0d5fbaaf64b9474e334e1"
Last-Modified: Tue, 22 Jan 2019 00:16:17 GMT
Vary: Accept-Encoding
Content-Range: bytes 1247595-13235121/13235122

127.0.0.1 - - [21/Jan/2019 16:18:31] "GET /untimely-pesky/download HTTP/1.1" 206 -
#1: 2.5 MiB, 19.63%
#1: 2.6 MiB, 20.40%
...
#1: 13.8 MiB, 109.38%
#1: 13.8 MiB, 109.43%
Stopped because transfer is complete
```

Only this time, it just keeps resuming until the download is complete.

So, what's the difference between the Tor Browser and wget requests? Why does Tor Browser keep failing? I really don't know, the http responses are pretty much exactly the same, only Tor Browser/Firefox appears to not like it and basically close the connection early, while wget thinks it's fine.

I'd like to get this into the 2.0 release if it's ready in time, but I think I'd rather not block the release with this, so if everything else is ready except this we can slate this for the following release, like 2.1.

Also, @heartsucker you can't run this locally because it segfaults on your laptop? That's not good! I can help you troubleshoot that problem so you can run it locally. Then you can test with actual Tor Browser. Basically just start download a share in Tor Browser, open the Downloads tab, and right-click on the download and click pause. Then to resume, right-click on it again and click resume. Right now, it resumes for like a second and then immediately fails, but the expected behavior is for it to download the rest of the file.

**Comment by heartsucker** [comment]
I am getting completely different behavior than you are. If I cancel then click retry, the download starts over.

First request for DL:

```http
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Encoding: gzip, deflate
Upgrade-Insecure-Requests: 1
Connection: keep-alive
Accept-Language: en-US,en;q=0.5
User-Agent: Mozilla/5.0 (Windows NT 6.1; rv:60.0) Gecko/20100101 Firefox/60.0
Host: orgtjorfjcsrbvlq.onion
```

First Response for DL:

```http
Content-Type: application/octet-stream
Content-Encoding: gzip
Content-Length: 50015287
Content-Disposition: attachment; filename=test-bytes.bin
Content-Security-Policy: default-src 'self'; style-src 'self'; script-src 'self'; img-src 'self' data:;
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Referrer-Policy: no-referrer
Server: OnionShare
Accept-Ranges: bytes
ETag: "sha256:ed868d9f4b4a00da707b82e1556cf44dff3d975933b4bfab7f02adcd932ce89d"
Last-Modified: Sun, 27 Jan 2019 16:14:00 GMT
Vary: Accept-Encoding
```

Second Request for DL:

```http
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Encoding: gzip, deflate
Connection: keep-alive
Accept-Language: en-US,en;q=0.5
User-Agent: Mozilla/5.0 (Windows NT 6.1; rv:60.0) Gecko/20100101 Firefox/60.0
Host: orgtjorfjcsrbvlq.onion
```

Note that here the `ETag` and `If-Unmodified-Since` headers aren't sent at all.

Second Response for DL:

```http
Content-Type: application/octet-stream
Content-Encoding: gzip
Content-Length: 50015287
Content-Disposition: attachment; filename=test-bytes.bin
Content-Security-Policy: default-src 'self'; style-src 'self'; script-src 'self'; img-src 'self' data:;
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Referrer-Policy: no-referrer
Server: OnionShare
Accept-Ranges: bytes
ETag: "sha256:ed868d9f4b4a00da707b82e1556cf44dff3d975933b4bfab7f02adcd932ce89d"
Last-Modified: Sun, 27 Jan 2019 16:14:00 GMT
Vary: Accept-Encoding
```

`torify wget --continue` works here and the server does the right thing.

System:

- Debian 9
- `torbrowser-launcher` 0.2.6-3.1
- Tor Browser 8.0.4

**Comment by heartsucker** [comment]
Ok this is absolutely a FF/TB issue, because if we copy the headers into curl and do

```bash
curl -H 'If-Match: "sha256:56d429ed15f4b07ab73a1e369602874639fc45c911685255fe35051c31155f1c"' \
    -H 'If-Unmodified-Since: Sun, 27 Jan 2019 17:26:16 GMT' \
    -H 'Connection: keep-alive' \
    -H 'Accept-Language: en-US,en;q=0.5' \
    -H 'Range: bytes=3588470-' \
    -H 'Host: 5sdujj4xna6na3n7.onion' \
    -H 'Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8' \
    -H 'Accept-Encoding: gzip, deflate' \
    -H 'User-Agent: Mozilla/5.0 (Windows NT 6.1; rv:60.0) Gecko/20100101 Firefox/60.0' \
    -vvv \
    http://5sdujj4xna6na3n7.onion/moocher-sulphuric/download
```

Everything works. This request is identical to the server as the one TB is sending. I have no idea what the problem is, and at this point I think we need a TB dev to point out what I'm doing wrong.

**Comment by sysrqb** [comment]
Finally got a chance to look at this. Sharing an Ubuntu ISO. It looks like there's an off-by-one, somewhere.

```
D/nsHttp Entity mismatch, expected 
'%22sha256:36d545e92dbcc5adeecd7b88d437d0b1e5ae630466f3b371104602eaac72b956%22/2074955820/Sat, 21 Sep 2019 21:36:51 GMT', got 
'%22sha256:36d545e92dbcc5adeecd7b88d437d0b1e5ae630466f3b371104602eaac72b956%22/2074955821/Sat, 21 Sep 2019 21:36:51 GMT', aborting [this=0x7f44f750a000]
```

First response:
```
HTTP/1.0 200 OK
Content-Type: application/x-iso9660-image
Content-Encoding: gzip
Content-Length: 2074955820
Content-Disposition: attachment; filename=ubuntu-19.04-desktop-amd64.iso
Content-Security-Policy: default-src 'self'; style-src 'self'; script-src 'self'; img-src 'self' data:;
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Referrer-Policy: no-referrer
Server: OnionShare
Accept-Ranges: bytes
ETag: "sha256:36d545e92dbcc5adeecd7b88d437d0b1e5ae630466f3b371104602eaac72b956"
```

Second request:
```
GET /conceal-arena/download HTTP/1.1
Host: 127.0.0.1:17638
User-Agent: Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:69.0) Gecko/20100101 Firefox/69.0
Accept: */*
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
Connection: keep-alive
Range: bytes=696846216-
If-Match: "sha256:36d545e92dbcc5adeecd7b88d437d0b1e5ae630466f3b371104602eaac72b956"
```

Second response:
```
HTTP/1.0 206 PARTIAL CONTENT
Content-Type: application/x-iso9660-image
Content-Encoding: gzip
Content-Length: 1378109604
Content-Disposition: attachment; filename=ubuntu-19.04-desktop-amd64.iso
Content-Security-Policy: default-src 'self'; style-src 'self'; script-src 'self'; img-src 'self' data:;
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Referrer-Policy: no-referrer
Server: OnionShare
Accept-Ranges: bytes
ETag: "sha256:36d545e92dbcc5adeecd7b88d437d0b1e5ae630466f3b371104602eaac72b956"
Last-Modified: Sat, 21 Sep 2019 21:36:51 GMT
Vary: Accept-Encoding
Content-Range: bytes 696846216-2074955820/2074955821
```

Firefox remembers the initial response said `Content-Length: 2074955820` and the second response (after pause/resume) said `Content-Range: bytes 696846216-2074955820/2074955821`. `curl` and `wget` don't have this memory, so they wouldn't fail in the same way.

**Comment by whew** [comment]
> It looks like there's an off-by-one, somewhere.

I can corroborate this. The `Content-Length` header is always one less than the length of the data that's actually sent.

The way the `Range` header is defined is really annoying and has probably caused issues like this many times in the past. If you have a file of  bytes then for any integers  and  where , the `Range: bytes=a-b` header includes exactly the bytes in the interval . I.e. all the bytes between  and , including both  and . Equivalent Python `range` notation would be `range(a, b + 1)`. The header is actually specifying  bytes, not , which would make way more sense.

I did some testing with a 1000-byte file to make it easy to see what's happening in each case:
| Range | True interval | True length | HTTP status code      | Content-Range     | Content-Length | Download size |
|-------|---------------|-------------|-----------------------|-------------------|----------------|---------------|
| None  |      | 100         | OK (200)              | None              | 99             | 100           |
| `0-`  |      | 100         | Partial Content (206) | None              | 99             | 100           |
| `0-9` |       | 10          | Partial Content (206) | `0-9/100`   | 9              | 10            |
| `10-` |     | 90          | Partial Content (206) | `10-99/100` | 89             | 90            |
| `-20` |     | 20          | Partial Content (206) | `80-99/100` | 19             | 20            |

In each case cURL actually discarded the extra byte at the end to be consistent with the `Content-Length` header.

**Comment by whew** [comment]
I did some more testing with pause/resume in Tor Browser. When you download the file without pausing and resuming then the whole file gets downloaded, even if `Content-Length` is wrong like it is here. If you do pause and resume, then it works iff the response isn't compressed:
|                          | `gzip` enabled | `gzip` disabled                  |
|--------------------------|----------------|----------------------------------|
| **`Content-Length` unchanged** | Fails          | Fails                            |
| **`Content-Length` +1**        | Fails          | Success, downloaded file matches |

**Comment by mikeperry-tor** [comment]
Doesn't onionshare zip the files typically? Or always? gzip should not really be that necessary in that case, right?

**Comment by SaptakS** [comment]
The code has also gone through a huge amount of refactoring and restructuring since this PR. Fixing the conflicts might be a nightmare. Might be easier to cherry pick the changes and then go ahead from there to see what are the issues still remaining. I am guessing a lot of the above issues might probably still be there.

**Comment by SaptakS** [comment]
@mikeperry-tor I have opened a new PR with the changes from this PR applied on the new code: https://github.com/micahflee/onionshare/pull/1321/files

cc: @micahflee @mig5 @whew

**Comment by SaptakS** [comment]
Closing this PR since https://github.com/onionshare/onionshare/pull/1321 was merged and we have a separate issue to track range requests error for Tor Browser

**Pull request closed without merging**